lib/mayaUsd/ufe/UsdLight.cpp

```diff
@@ -0,0 +1,499 @@
+//
+// Copyright 2022 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdLight.h"
+
+#include "private/Utils.h"
+
+#include <mayaUsd/ufe/Utils.h>
+#include <mayaUsd/utils/util.h>
+
+#include <pxr/usd/usdLux/distantLight.h>
+#if PXR_VERSION < 2111
+#include <pxr/usd/usdLux/light.h>
+#else
+#include <pxr/usd/usdLux/lightAPI.h>
+#endif
+#include <pxr/usd/usdLux/rectLight.h>
+#include <pxr/usd/usdLux/shadowAPI.h>
+#include <pxr/usd/usdLux/shapingAPI.h>
+#include <pxr/usd/usdLux/sphereLight.h>
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+template <typename ValueTypeIn>
+class SetValueUndoableCommandImpl : public Ufe::SetValueUndoableCommand<ValueTypeIn>
```

Comment:
Unfortunately we can't implement undo / redo in USD by setting the previous value on undo, because this is incomplete.  If you imagine a situation where the edit target is set to a higher-priority layer which has no current opinion, undo should completely remove the opinion, NOT set the new opinion to the previous value.

Fortunately we have helper classes for this which make it easy to do the right thing.  See the existing
https://github.com/Autodesk/maya-usd/blob/77c5b01b67a63b5c280bec4e1c8a30472c8431e2/lib/mayaUsd/ufe/UsdUndoableCommand.h#L36
and
https://github.com/Autodesk/maya-usd/blob/6e1c88af1bfb3652c7f6999aea41847c14fd1fee/lib/mayaUsd/ufe/UsdAttribute.cpp#L199
and their uses in the code base, and pick the one that's right for you.